RecoEgamma/EgammaIsolationAlgos/plugins/isolation_cones/PhotonPFIsolationWithMapBasedVeto.cc

```diff
@@ -0,0 +1,170 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+
+#include "DataFormats/Candidate/interface/Candidate.h"
+#include "DataFormats/Candidate/interface/CandidateFwd.h"
+
+#include "DataFormats/EgammaCandidates/interface/Photon.h"
+#include "DataFormats/EgammaCandidates/interface/PhotonFwd.h"
+
+#include "DataFormats/PatCandidates/interface/Photon.h"
+
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+
+#include "PhysicsTools/IsolationAlgos/interface/IsoDepositVetoFactory.h"
+#include "DataFormats/RecoCandidate/interface/IsoDepositVetos.h"
+
+#include "PhysicsTools/IsolationAlgos/interface/CITKIsolationConeDefinitionBase.h"
+#include "DataFormats/Math/interface/deltaR.h"
+
+#include "DataFormats/EcalDetId/interface/EcalSubdetector.h"
+
+
+#include <unordered_map>
+
+namespace reco {
+  typedef edm::Ptr<reco::Photon> recoPhotonPtr;
+}
+
+namespace pat {
+  typedef edm::Ptr<pat::Photon> patPhotonPtr;
+  typedef edm::Ptr<pat::PackedCandidate> PackedCandidatePtr;
+}
+
+// This template function finds whether theCandidate is in thefootprint
+// collection. It is templated to be able to handle both reco and pat
+// photons (from AOD and miniAOD, respectively).
+template <class T, class U>
+bool isInFootprint(const T& thefootprint, const U& theCandidate) 
+{
+    for ( auto itr = thefootprint.begin(); itr != thefootprint.end(); ++itr ) 
+    {
+      if( itr->key() == theCandidate.key() ) return true;// This line is commented out, because it's itroducing trouble for pfNoPileUpCandidates
+     }
+    return false;
+}
+
+//This function is needed because pfNoPileUpCandidates have changed keys, 
+//and thus the solution is to use sourceCandidatePtr(0) 
+// This function *shouldn't be used for packedCandidate*
+template <class T, class U>
+bool isInFootprintAlternative(const T& thefootprint, const U& theCandidate) 
```

Comment:
On 2/12/16 4:30 AM, Ivan Shvetsov wrote:
> In
> RecoEgamma/EgammaIsolationAlgos/plugins/isolation_cones/PhotonPFIsolationWithMapBasedVeto.cc
> <https://github.com/cms-sw/cmssw/pull/13223#discussion_r52735717>:
> 
>> +// photons (from AOD and miniAOD, respectively).
>> +template <class T, class U>
>> +bool isInFootprint(const T& thefootprint, const U& theCandidate) 
>> +{
>> +    for ( auto itr = thefootprint.begin(); itr != thefootprint.end(); ++itr ) 
>> +    {
>> +      if( itr->key() == theCandidate.key() ) return true;// This line is commented out, because it's itroducing trouble for pfNoPileUpCandidates
>> +     }
>> +    return false;
>> +}
>> +
>> +//This function is needed because pfNoPileUpCandidates have changed keys, 
>> +//and thus the solution is to use sourceCandidatePtr(0) 
>> +// This function *shouldn't be used for packedCandidate*
>> +template <class T, class U>
>> +bool isInFootprintAlternative(const T& thefootprint, const U& theCandidate) 
> 
> I'm not sure what you mean by "partially instantiate isInFootprint for
> PFCandidate".


Sorry, the proper verb is "specialize" (not instantiate)

http://en.cppreference.com/w/cpp/language/partial_specialization

It looks like the actual type "U" is different and is known for both cases



> The reason for using 2 functions is that for AOD we want to use
> pfNoPileUpCandidates, this breaks keys and one has to look for
> sourceCandidate to get key properly.
> 
> I was thinking that doing it from sourceCandidate would be more generic,
> but I have checked it for miniAOD and
> theCandidate->sourceCandidate(0).key() gives different result than
> theCandidate.key()
> 
> Does this make sense to you?
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/13223/files#r52735717>.
>